### ISSUE TITLE
A suspicious implicit ambiguity with higher types parameterized with an invariant type.

### ISSUE THREAD

**Issue by noresttherein** [scala/bug#12238]
## reproduction steps

Ok, maybe this is intentional, a 'feature', or 'wontfix', but still an unpleasant occurence:

using Scala 2.13.4,

```scala

	class Maybe[+T](val opt :Option[T])

	class MaybeNo {
		implicit val maybeNo :Maybe[Nothing] = new Maybe(None)
	}

	object Maybe extends MaybeNo {
		implicit def maybeYes[T](implicit ev :T) :Maybe[T] = new Maybe[T](Some(ev))
	}

	class Invariant[T]

	object Invariant {
		implicit def any[T] = new Invariant[T]
	}


	def test[T] = fff__[T]   // <- here!
	def fff__[T](implicit ev :Maybe[Invariant[T]]) = ev
```

## problem
```
ambiguous implicit values:
 both value maybeNo in class MaybeNo of type Playground.Maybe[Nothing]
 and method maybeYes in object Maybe of type [T](implicit ev: T): Playground.Maybe[T]
 match expected type Playground.Maybe[Playground.Invariant[T]]
	implicit def test[T] = fff__[T]
```
For one, `Maybe[Nothing]` doesn't even conform to `Maybe[Invariant[T]]`.
For two (second?), if `maybeYes` conforms to the expected type, then `maybeNo` shouldn't even be considered, should it?

**Comment by SethTisue** [comment]
> Maybe[Nothing] doesn't even conform to Maybe[Invariant[T]].

it doesn't? why not?

**Comment by noresttherein** [comment]
Ok, that was a stupid claim on my part, I have no idea how it went out of my mouth.

**Comment by SethTisue** [comment]
no worries, we've all been there :-)

we appreciate your bug reports, but I'm afraid I have to ask you: if you aren't sure whether something is a bug or not, please open a thread on https://users.scala-lang.org first, and then only come here and open a bug report once you have become reasonably certain that the root cause is in fact a bug in Scala

the other thing we ask is that bug reports be minimized. we typically accept un-minimized bug reports from first-time reporters, but you've opened a bunch of tickets now, so I think it's fair to ask that you also be reasonably certain, before filing the bug, that what you have is _absolutely_ the minimal reproduction to the best of your ability. folks on Discourse (perhaps me!) can help with this if you're not sure.

just looking at the code above, it certainly doesn't look minimal to me. it's just a gut feeling, but you've now filed a string of not-especially-minimized bug reports, so the suspicion seems reasonable

**Comment by som-snytt** [comment]
Maybe `Maybe[Nothing]` is more specific than `Maybe[T]` but `maybeNo` is in superclass relative to `maybeYes`.

I don't get why the space is before the colon instead of after it. Maybe there is a Scalafix. :)

Also a Scalafix to move companions closer.

**Comment by noresttherein** [comment]
> Maybe `Maybe[Nothing]` is more specific than `Maybe[T]` but `maybeNo` is in superclass relative to `maybeYes`.
Yeah, I get that. It was really a moment of dumbing out after a couple of days trying to make a collection of a hundered implicits behave.

> I don't get why the space is before the colon instead of after it. Maybe there is a Scalafix. :)
Because that's how my fingers where trained in Pascal a quarter of a century ago. IntelliJ has a setting for this so I can't be alone. I also use tabs (as a compromise in 2vs4 spaces debate) - is that also not welcome here?

> Also a Scalafix to move companions closer. Edit: :) 
And obfuscate implicit precedence order in a type class? Extended classes with 'worse' implicits are in general much shorter than the objects they extend, so `maybeNo` in any more complex case would be hidden. Maybe it's a cultural thing, a preference of having the super class at the top and its specialisations following down the line rather than the other way round.

>I wonder if it's symptomatic here of how the companion is assembled. I wonder if underscore has a how-to, depending on your goals?
Didn't even understand this one.

And acknowledged, I won't post reports in the future unless I minimise them to the best of my ability.

**Comment by SethTisue** [comment]
I don't think anyone cares about tabs/spaces in this context, especially since we're just reading the code (or copying/pasting it) and not maintaining it.

To the best of my recollection, I have never seen _any_ other Scala code written in the `x :Int` style — so actually it is jarring enough to impair readability for your audience here.

**Comment by SethTisue** [comment]
> Didn't even understand this one

Attempting to puzzle out some of Som's more gnomic utterances is a fairly common pastime around here :-)

**Comment by som-snytt** [comment]
> tabs

I used tabs when I wrote C code and saved files on floppies. I use them now in the office because of the cultural affiliation of the other guy. I try to embrace diversity.

> 2vs4 spaces debate

but let's be clear that there is no debate. :)

> Pascal

Thanks for explaining. I've forgotten my brief Pascal from college.

> gnomic 

Didn't intend, this time, to be a gnome or troll. I meant to say Underscore consulting might cover how to prioritize implicits in their books. I follow the conversations, but I have never prioritized implicits "in anger."

**Comment by joroKr21** [comment]
> Maybe Maybe[Nothing] is more specific than Maybe[T] but maybeNo is in superclass relative to maybeYes.

Precisely, and this is intentional. Unfortunately providing an implicit `Maybe[Nothing]` for a covariant `Maybe` will always be the most specific instance in implicit scope (unless you have subclasses of `Maybe`) so it probably defeats the purpose.

**Comment by noresttherein** [comment]
> > Maybe Maybe[Nothing] is more specific than Maybe[T] but maybeNo is in superclass relative to maybeYes.
> Precisely, and this is intentional. 

Yes, the moment Seth posted I realised how dumb I was in that moment. If I could close the ticket, I would have, but now let it  hang here as a lesson of humility.

> but let's be clear that there is no debate. :)
Every company I worked for professionally since 2013 used 4 spaces and not because of me, so it depends on your bubble.
It somewhat depends on font and colour settings, but I have no doubts whatsover that 4 spaces can be scientifically proven to be better from the cognitive point of view. But then, there are coding styles which have weight of their own, and one would no longer be in the cool academic club without 'pure' 2 spaces. They make all code 50% more functional, it is known ;)

> > Pascal
> 
> Thanks for explaining. I've forgotten my brief Pascal from college.
I also wanted to say ML/OCaml, but after a brief looking online I see that most code uses spaces on both sides, so it might have been a local thing (like, one TA in their teaching materials wrote like that and all his students followed). It certainly wasn't me being a hipster.

> Didn't intend, this time, to be a gnome or troll. I meant to say Underscore consulting might cover how to prioritize implicits in their books. I follow the conversations, but I have never prioritized implicits "in anger."

Ah, as Underscore was the first word of the sentence, i didn't realise it's a proper name and wondered if you were speaking about `ffff__`.  The thing is, I know how implicits are proritised and even have a good hunch why the type inferer sometimes ignores some. But I did run into two legitimate bugs in that area recently and was burnt out by somewhat tedious refactors trying to understand, isolate and work around them. So I guess my mind just gave up at that moment. And I wasn't in anger, if that's what you mean; I am sorry if I came out that way.
